File: Lambdas/productEDM/.debug/lambda_functions/registerProduct.py
"""
    Author: Hieu Phung; SOLERS INC
    Contact: hieu.phung@solers.com
    Last modified: 2019-01-02
    Python Version: 3.6
"""

import os
import math
import json
import elasticsearch
import certifi
import psycopg2

from ndeutil import *


print('Loading registerProduct')

def lambda_handler(event, context):
    print(event)
    
    # get the msg body for proxy integration case
    if 'resource' in event:
        if event.get('body') is None:
            return createErrorResponse(400, "Validation error", 
                "body (product definition json) required")
        if isinstance(event.get('body'), dict):
            inputBody = event.get('body')
        else:
            inputBody = json.loads(event.get('body'))
    else:
        inputBody = event
    
    productDesc = inputBody.get('ProductDefinition')
    print("productDesc", productDesc)
    
    # did the client request a create or update
    isUpdate = False
    if 'resource' in event and event['resource'] == '/product/{productid}':
        isUpdate = True
        
    print('isUpdate?', isUpdate)
    
    try:
        conn = psycopg2.connect(
                host = os.environ['RD_HOST'],
                dbname = os.environ['RD_DBNM'],
                user = os.environ['RD_USER'],
                password = os.environ['RD_PSWD']
                )
        
        # check for existing
        cur = conn.cursor()
        
        if 'Platforms' in productDesc:
            cur.execute("SELECT platformid FROM platform WHERE platformname = %s", 
                            (productDesc.get('Platforms')[0].get('Platformname'),) )
            res = cur.fetchone()
            if res is not None: 
                platformId = res[0]
                print("platformId", platformId)
                cur.execute("SELECT pd.productid FROM productdescription pd, productplatform_xref prx "
                    "WHERE pd.productshortname = %s AND pd.productid = prx.productid AND prx.platformid = %s", 
                    (productDesc['PRODUCTSHORTNAME'], platformId,) )
            else:
                print("WARN: No Platform registered with name:", productDesc['Platforms'][0]['Platformname'])
                cur.execute("SELECT productid FROM productdescription WHERE productshortname = %s", 
                (productDesc['PRODUCTSHORTNAME'],) )
        else:
            cur.execute("SELECT productid FROM productdescription WHERE productshortname = %s", 
                (productDesc['PRODUCTSHORTNAME'],) )
                
        res = cur.fetchone()
        productId = res[0] if res is not None else None
        cur.close()
        
        print("productId", productId)
        
        # if isUpdate and {id} is None, proceed to register normally
        if not isUpdate and productId is not None:
            s = "Product already exists with short name: " + productDesc['PRODUCTSHORTNAME']
            print(s)
            return createErrorResponse(409, "Conflict: resource exists", s)
        elif isUpdate:
            if productId is None:
                s = "Unable to update; product with this short name NOT registered: " + productDesc['PRODUCTSHORTNAME']
                print(s)
                return createErrorResponse(404, "Not Found", s)
            elif productId != int(event['pathParameters']['productid']):
                s = "Unable to update; requested URI {productid} does not match existing Product ID: " + str(productId)
                print(s)
                return createErrorResponse(400, "Validation error", s)
        
        # start processing the input, rollback if any sections fail
        try:
            productId = doProductDesc(conn, productDesc, productId, isUpdate)
        except ValueError as e:
            print("Exception in doProductDesc:", e)
            return createErrorResponse(400, "Validation error", str(e))
            
        print("productId", productId)
        
        if 'ProducQualitySummarys' in productDesc:
            print(productDesc['ProductQualitySummarys'])
            try:
                doProductQualitySummarys(conn, productDesc['ProductQualitySummarys'], productId, isUpdate)
            except ValueError as e:
                print("Exception in doProductQualitySummarys:", e)
                return createErrorResponse(400, "Validation error", str(e))
        
        if 'IngestProcessSteps' in productDesc:
            print(productDesc['IngestProcessSteps'])
            try:
                doIngestProcessSteps(conn, productDesc['IngestProcessSteps'], productId, isUpdate)
            except ValueError as e:
                print("Exception in doIngestProcessSteps:", e)
                return createErrorResponse(400, "Validation error", str(e))
                
        if 'HOSTNAME' in productDesc:
            try:
                doProductDataSource(conn, productDesc, productId, isUpdate)
            except ValueError as e:
                print("Exception in doProductDataSource:", e)
                return createErrorResponse(400, "Validation error", str(e))
        
        if 'Platforms' in productDesc:
            try:
                doPlatforms(conn, productDesc['Platforms'], productId, isUpdate)
            except ValueError as e:
                print("Exception in doPlatforms:", e)
                return createErrorResponse(400, "Validation error", str(e))
        
        # Save the entire JSON doc (including ProductDescription root) to ES
        esSuccess = createElasticDoc(productId, json.dumps({"ProductDefinition": productDesc}))
        
        if esSuccess:
            conn.commit()
        else:
            conn.rollback()
            return createErrorResponse(500, "Error encountered updating ElasticSearch", str(e))
        
        response = {
            "isBase64Encoded": False,
            "statusCode": 200 if isUpdate else 201,
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": '*'
            },
            "body": json.dumps({
                "message": "Product was " + ("updated" if isUpdate else "registered") + " successfully",
                "productId": productId
            })
        }
        return response
    except Exception as e:
        print("ERROR: " + str(e))
        return createErrorResponse(500, "Internal error", "An exception was encountered, please try again later")
    finally:
        conn.rollback()
        conn.close()


def doProductDesc(conn, productDesc, productId, isUpdate):
    print(">> Starting doProductDesc")
    
    if productDesc.get('PRODUCTSTATUS') == "": productDesc['PRODUCTSTATUS'] = 'NOP'
    if productDesc.get('PRODUCTSTATUS') not in ['OP','TEST','NOP']:
        raise ValueError('PRODUCTSTATUS value is invalid. Valid values: [OP, TEST, NOP]')
    
    if productDesc.get('PRODUCTARCHIVE') == "": productDesc['PRODUCTARCHIVE'] = '0'
    if productDesc.get('PRODUCTARCHIVE') not in ['0','1']:
        raise ValueError('PRODUCTARCHIVE value is invalid. Valid values: [0, 1]')
    
    if "PRODUCTDISTFLAG" not in productDesc or productDesc.get('PRODUCTDISTFLAG') not in ['TRUE','FALSE']:
        productDesc['PRODUCTDISTFLAG'] = 'TRUE'
    
    print(productDesc.get('PRODUCTCOVERAGEGAPINTERVAL_DS'))
    if isStrEmptyNull(productDesc.get('PRODUCTCOVERAGEGAPINTERVAL_DS')):
        productDesc['PRODUCTCOVERAGEGAPINTERVAL_DS'] = "PT0S" #ISO8601 duration format for 0 sec
    
    if productDesc.get('PRODUCTCIPPRIORITY') == "":
        productDesc['PRODUCTCIPPRIORITY'] = None
    
    if productDesc.get('PRODUCTAVAILABILITYDATE') == "":
        productDesc['PRODUCTAVAILABILITYDATE'] = None
    
    if productDesc.get('PRODUCTSPATIALAREA') == "":
        productDesc['PRODUCTSPATIALAREA'] = None
            
    productDesc['PRODUCTESTAVGFILESIZE'] = math.ceil(float(productDesc.get('PRODUCTESTAVGFILESIZE')))
    
    cur = conn.cursor()
    try:
        cur.execute("SELECT ingestincomingdirectoryid FROM ingestincomingdirectory "
            "WHERE ingestdirectoryname = %s", (productDesc['INGESTDIRECTORYNAME'],) )
        ingestDirId = cur.fetchone()[0]
    except Exception as e:
        print('ERROR: Invalid INGESTDIRECTORYNAME:', productDesc['INGESTDIRECTORYNAME'])
        raise ValueError('Invalid INGESTDIRECTORYNAME')
    
    productGroupId = None
    try:
        if productDesc['PRODUCTGROUPNAME'] != "" and productDesc['PRODUCTGROUPNAME'] != "null":
            cur.execute("SELECT productgroupid FROM productgroup WHERE productgroupname = %s", 
                (productDesc['PRODUCTGROUPNAME'],) )
            productGroupId = cur.fetchone()[0]
    except Exception as e:
        print('WARN: No product group with name:', productDesc['PRODUCTGROUPNAME'])
        
    if not isUpdate:
        cur.execute("INSERT INTO productdescription (productid) "
            "VALUES (nextval('s_productdescription\')) RETURNING productid")
        productId = cur.fetchone()[0]
        print("new productId", productId)
        if productDesc['PRODUCTFILENAMEEXTENSION'] == "h5":
            cur.execute("INSERT INTO hdf5_structure values (%s)", (productId,) )
        elif productDesc['PRODUCTFILENAMEEXTENSION'] == "nc":
            cur.execute("INSERT INTO netcdf4_structure values (%s)", (productId,) )
            
    excludeFields = ['ProductQualitySummarys', 'IngestProcessSteps', 'Platforms',
        'INGESTDIRECTORYNAME', 'PRODUCTGROUPNAME','HOSTNAME','PRODUCTPROVIDERDIRECTORY',
        'PRODUCTPOLLINGFREQUENCY','PRODUCTPROVIDERPOLLINGRETRIES', 'PRODUCTSPATIALAREA']
    fieldsSnippet = ', '.join([f + "=%(" + f + ")s" for f in list(productDesc.keys()) if f not in excludeFields])
    print(fieldsSnippet)
    
    cur.execute("UPDATE productdescription SET " + fieldsSnippet + " WHERE productid = %(productid)s",
        {'productid': productId, 'INGESTINCOMINGDIRECTORYID': ingestDirId, 'PRODUCTGROUPID': productGroupId, **productDesc})
        
    cur.close()
    
    print(">> Finished doProductDesc")
    return productId
    

def doProductQualitySummarys(conn, pqsObj, productId, isUpdate):
    print(">> Starting doProductQualitySummarys")
    
    cur = conn.cursor()
    for pqs in pqsObj:
        doInsert = True
        if isUpdate:
            cur.execute("SELECT productqualitysummaryname FROM productqualitysummary WHERE productid = %s "
            "AND productqualitysummaryname = %s", (productId, pqs['PRODUCTQUALITYSUMMARYNAME'],))
            row = cur.fetchone()
            if row is not None:
                doInsert = False
        if doInsert:
            cur.execute("INSERT INTO productqualitysummary VALUES (%s, %s, %s, %s)", (productId, 
                pqs['PRODUCTQUALITYSUMMARYNAME'], pqs['PRODUCTQUALITYSUMMARYTYPE'], pqs['PRODUCTQUALITYDESCRIPTION'],))
            print("Added PQS: ", pqs['PRODUCTQUALITYSUMMARYNAME'])
    
    cur.close()
    print(">> Finished doProductQualitySummarys")


def doIngestProcessSteps(conn, ipsObj, productId, isUpdate):
    print(">> Starting doIngestProcessSteps")
    
    cur = conn.cursor()
    if isUpdate:
        cur.execute("DELETE FROM ingestprocessstep WHERE productID = %s", (productId,))
    
    for ips in ipsObj:
        if (ips['IPSFAILSINGEST'] == "" or ips['IPSENABLE'] == "" or ips['IPSENABLE'] == ""):
            raise ValueError('Missing IPSFAILSINGEST or IPSENABLE or IPSENABLE for IPS', ips['NSFDESCRIPTION'])
        
        cur.execute("SELECT nsfid FROM nde_supportfunction WHERE nsfdescription = %s", (ips['NSFDESCRIPTION'],))
        row = cur.fetchone()
        if row is None: 
            raise ValueError('Unregistered NSFDescription:', ips['NSFDESCRIPTION'])
        
        cur.execute("INSERT INTO ingestprocessstep (productid, nsfid, ipsoptionalparameters, ipsfailsingest, "
            "ipsenable, ipsdoretransmit) VALUES (%s, %s, %s, %s, %s, %s)", (productId, row[0],
            ips['IPSOPTIONALPARAMETERS'], ips['IPSFAILSINGEST'], ips['IPSENABLE'], ips['IPSDORETRANSMIT']))
            
        print("Added IPS: ", ips['NSFDESCRIPTION'])
    
    cur.close()
    print(">> Finished doIngestProcessSteps")
    


def doProductDataSource(conn, productDesc, productId, isUpdate):
    print(">> Starting doProductDataSource")
    
    cur = conn.cursor()
    try:
        cur.execute("SELECT hostid FROM externaldatahost WHERE hostname = %s", (productDesc['HOSTNAME'],))
        hostId = cur.fetchone()[0]
    except Exception as e:
        print('ERROR: External Host not registered:', productDesc['HOSTNAME'])
        raise ValueError('External Host not registered')
    
    if isUpdate:
        cur.execute("DELETE FROM externaldatahost WHERE productid = %s AND hostid = %s", (productId, hostId))
    
    cur.execute("INSERT INTO productdatasource (productid, hostid, productproviderdirectory, productpollingfrequency, "
        "productproviderpollingretries ) values ( %s, %s, %s, %s, %s)", (productId, hostId, 
        productDesc['PRODUCTPROVIDERDIRECTORY'], productDesc['PRODUCTPOLLINGFREQUENCY'], 
        productDesc['PRODUCTPROVIDERPOLLINGRETRIES']))
    
    cur.close()
    
    print(">> Finished doProductDataSource")
    

def doPlatforms(conn, platformObj, productId, isUpdate):
    print(">> Starting doPlatforms")
    
    cur = conn.cursor()
    if isUpdate:
        cur.execute("DELETE FROM productplatform_xref WHERE productid = %s", (productId,))
    
    for platform in platformObj:
        cur.execute("SELECT platformid FROM platform WHERE platformname = %s", (platform['Platformname'],) )
        row = cur.fetchone()
        if row is not None: 
            platformId = row[0]
            cur.execute("INSERT INTO productplatform_xref VALUES ( %s, %s )", (platformId, productId));
            print("Added Platform XREF: ", platformId, "<=>", productId)
        else:
            print("ERROR: No Platform registered with name: " + platform['Platformname'])
            raise ValueError("No Platform registered with name: " + platform['Platformname'])
    
    cur.close()
    
    print(">> Finished doPlatforms")


def createElasticDoc(productId, productJson):
    print(">> Starting createElasticDoc")
    
    es = elasticsearch.Elasticsearch(
        [os.environ['ES_HOST']],
        port = os.environ['ES_PORT'],
        use_ssl = True,
        verify_certs = True,
        ca_certs = certifi.where()
        )
    
    try:
        res = es.index(
            index = "product", 
            doc_type = '_doc', 
            id = productId, 
            body = productJson
        )
        
        if res['result'] == 'created' and res['_id'] == str(productId):
            print("Successfully added to Index:", res)
            return True
        elif res['result'] == 'updated' and res['_id'] == str(productId):
            print("Successfully updated Index:", res)
            return True
        else:
            print("Failed to add to Index:", res)
            return False
    except Exception as e:
        print("ERROR: Failed to add to Elastic Search:", str(e))
        return False
    finally: 
        print(">> Finished createElasticDoc")
    

File: Tools/perl/pds/registerProduct.pl
#!/usr/bin/env perl

#
# name: registerProduct.pl
# revised: 20111220 lhf, add xsd validation, update capability
#          20090225 tjf make work w/ingestdirectoryname string
#          20090615 tjf changed productmetadataxml to productprofilelink, no xml type
#                     columns are handled by this script
#          20091028 lhf added PRODUCTGROUPNAME to xml and section for PRODUCTGROUPID
#          20091127 lhf update to include PRODUCTDATASOURCE information
#          20091208 lhf add structure row
#          20100511 lhf added PRODUCTCOVERAGEGAPINTERVAL_DS
#          2010l203 lhf allows null in PRODUCTCOVERAGEGAPINTERVAL_DS
#          20110111 htp update to include Ingest Process Step information
#          20110111 htp update to rollback DB inserts if one fails
#          20110408 lhf RENTETIONPERIODHR -> RETENTIONPERIODHR
#          20110422 htp removed hardcoded IPS flags and optional params
#          20110811 htp general refactoring, more prints, use $dbh->begin_work
#                        added warning for prod home dir
#          20120522 lhf build 4/5 updates
#          20120530 lhf PRODUCTPLATFORM_XREF info added
#          20121102 lhf correct for PRODUCTOBSTIMEPATTERN and PRODUCTIRMESSAGEEXTENSION
#          20121212 lhf add logs/pds
#          20130123 lhf validate PRODUCTSTATUS, default to NOP
#		   20131205 dcp modified to allow update of product quality summary table
#		   20141013 dcp, added platform mapping to product IDs
#	       20141118 teh remove product homedir=shortname test.  OBE with GVF_IP products especially.
#          20171120 teh add handling for new optional PRODUCTDISTFLAG field. ENTR-4319
#	20180308 pgm postgres

use XML::Simple;
use XML::LibXML;
use Data::Dumper;
use DBI;

use Getopt::Std;
getopt('m:f:');

# Check for comand line args
if ( ! $opt_m || ! $opt_f) {
        print "\nUsage $0 -m <mode> -f <XML Filename>\n\n";
        exit(1);
}

chomp( $YMDHMS = `date +%Y%m%d_%H%M%S` );
my $rfn = reverse( $opt_f );
if ( index(reverse($opt_f),"/") > 0) {
  $rfn = substr(reverse($opt_f),0,index(reverse($opt_f),"/"));
}

my $logfile = "/opt/data/nde/" . $opt_m . "/logs/pds/PRODUCTDESCRIPTION_" . reverse($rfn) . "_" . $YMDHMS . ".log";
open ( LOG, ">$logfile" ) or
  die "ERROR Error opening logfile, ($logfile) exiting...\n";

logger( "INFO Execution started..." );

my $schema = XML::LibXML::Schema->new(location => "/opt/data/nde/$opt_m\/xsds/PRODUCTDESCRIPTION.xsd");
my $parser = XML::LibXML->new;
my $doc = $parser->parse_file( $opt_f );

eval {
    $schema->validate($doc);
};

if ($@) {
    logger( "ERROR $opt_f file FAILED validation: $@" ) if $@;
    exit( 1 );
}
if (-r "../common/ndeUtilities.pl") {
  require "../common/ndeUtilities.pl";
} else {
  require "/opt/apps/nde/$opt_m/common/ndeUtilities.pl";
}
my $pw = promptForPassword("\n  Enter $opt_m password: ");

# postgres is case-sensitive convert $opt_m to lowercase 

my $dbh = DBI->connect( "dbi:Pg:dbname=nde_dev1;host=nde-pgsql.cixifosqwtgg.us-east-1.rds.amazonaws.com", lc($opt_m), 'nde' )
          or die "\nDatabase Connection Error: $DBI::errstr\n";
#$dbh->{LongReadLen}=64000;
$dbh->begin_work;
my $update = "n";

my %ref;
my $ref = XMLin( $opt_f, forcearray => ['ProductQualitySummary','IngestProcessStep','Platform'] );

#logger( Dumper( $ref ) );


### Check for existing product
my $psn = $ref->{PRODUCTSHORTNAME};

my $platformName = $ref->{Platforms}{Platform}[0]->{Platformname};
logger( "INFO $platformName" );
$sth = runSQL( $dbh, "select PLATFORMID from PLATFORM where PLATFORMNAME = '$platformName'" );
my ( $platformID ) = $sth->fetchrow_array();

if (!defined($platformID)) {
	$sth = runSQL($dbh, "select PRODUCTID from PRODUCTDESCRIPTION where PRODUCTSHORTNAME = '$psn'" );
} else {
	$sth = runSQL($dbh, "select pd.PRODUCTID from PRODUCTDESCRIPTION pd, PRODUCTPLATFORM_XREF prx where PRODUCTSHORTNAME = '$psn'
							and pd.PRODUCTID = prx.PRODUCTID and PLATFORMID = $platformID" );
}
my ($productID) = eval { $sth->fetchrow_array()};
if ($productID ne "" ) {
  print "WARNING Product $psn already exists (ID: $productID), continue with update? (y/n): ";
  chomp($update = <STDIN>);
  if (($update ne 'y') && ($update ne 'n')) {
		logger( "ERROR  You did not enter a \"y\" or a \"n\". Exiting...");
		$dbh->rollback();
		exit 1;
	} elsif ($update eq "n") {
		logger( "INFO  Not updating product $psn and exiting...");
		$dbh->rollback();
		exit 1;
	}
}

## Check product home dir
#if( $ref->{PRODUCTHOMEDIRECTORY} ne ("products/" . $psn) ){
#        print "WARNING Product home dir ($ref->{PRODUCTHOMEDIRECTORY}) does not match product short name, continue? (y/n) :";
#        if( <> eq "n\n"){       print "\nExiting\n\n"; $dbh->rollback; exit 1; }
#}

# Validate status
my $productStatus = $ref->{PRODUCTSTATUS};
$productStatus = "NOP", if( $productStatus eq "" );
if( ! grep(/$productStatus/, ('OP','TEST','NOP') ) ){
        print "WARNING Product Status value invalid ($productStatus), valid values are OP, TEST, and NOP\n";
        print "\nExiting\n\n"; $dbh->rollback; exit 1; }
        
# Validate archive
my $productArchive = $ref->{PRODUCTARCHIVE};
$productArchive = "0", if ( $productArchive eq "" );
if( ! grep(/$productArchive/, ('0','1') ) ){
        print "WARNING Product Archive value invalid ($productArchive), valid values are 0 or 1\n";
        print "\nExiting\n\n"; $dbh->rollback; exit 1; }

# Get some required values
print "  Looking up IncomingDirID and ProductGroupID ...";

my $sql= "select INGESTINCOMINGDIRECTORYID from INGESTINCOMINGDIRECTORY where INGESTDIRECTORYNAME = '$ref->{INGESTDIRECTORYNAME}'" ;
$sth = runSQL($dbh, $sql );
$sth->execute;
my ( $incomingDirID ) = $sth->fetchrow_array();
if ($incomingDirID eq "" ) {
        print "    could not find ID for Ingest Incoming Directory \"$ref->{INGESTDIRECTORYNAME}\"\n";
        $dbh->rollback; exit( 1 );
}
my $productGroupID = "null";
if ( $ref->{PRODUCTGROUPNAME} ne "null" && $ref->{PRODUCTGROUPNAME} ne "" )  {
  $sth = runSQL($dbh, "select PRODUCTGROUPID from PRODUCTGROUP where PRODUCTGROUPNAME = '$ref->{PRODUCTGROUPNAME}'");
  ( $productGroupID ) = $sth->fetchrow_array();
}
print " done\n";

my $productDistFlag = "TRUE";
if ( $ref->{PRODUCTDISTFLAG} ne "" ) {
  	$productDistFlag = uc $ref->{PRODUCTDISTFLAG};
  	#There are only 2 allowable values now, after filtering out null and uppercasing: TRUE or FALSE 
  	#Enforce:
  	if ($productDistFlag ne "FALSE" ) {$productDistFlag = "TRUE";}
}


my $pcgi_ds;
if ( $ref->{PRODUCTCOVERAGEGAPINTERVAL_DS} eq "" || $ref->{PRODUCTCOVERAGEGAPINTERVAL_DS} eq "null" ) {
  $pcgi_ds = "interval '0' second";
  }
else {
  $pcgi_ds = $ref->{PRODUCTCOVERAGEGAPINTERVAL_DS};
}


# Register product
# if insert, then store stub (psn may not be updated)
if ( $update eq "n" ) {
  logger( "INFO New product: $psn ..." );
  $sql = "insert into ProductDescription (PRODUCTID, PRODUCTSHORTNAME)
           values (nextval('S_PRODUCTDESCRIPTION'), '$psn')";
  $sth = runSQL( $dbh, $sql );

  $sth = runSQL($dbh, "select PRODUCTID from PRODUCTDESCRIPTION where PRODUCTSHORTNAME = '$psn'
								and PRODUCTFILENAMEPATTERN IS NULL" );
  $productID = eval { $sth->fetchrow_array()};
}
logger( "INFO Completing registration of product: $psn (ProductId: $productID) ..." );
$sql = "update PRODUCTDESCRIPTION set PRODUCTSHORTNAME='$psn', PRODUCTLONGNAME='$ref->{PRODUCTLONGNAME}',
         PRODUCTDESCRIPTION='$ref->{PRODUCTDESCRIPTION}', PRODUCTMETADATALINK='$ref->{PRODUCTMETADATALINK}',
          PRODUCTPROFILELINK='$ref->{PRODUCTPROFILELINK}', PRODUCTTYPE='$ref->{PRODUCTTYPE}', PRODUCTFILENAMEPREFIX='$ref->{PRODUCTFILENAMEPREFIX}',
           PRODUCTIDPSMNEMONIC='$ref->{PRODUCTIDPSMNEMONIC}', PRODUCTAVAILABILITYDATE='$ref->{PRODUCTAVAILABILITYDATE}',
            PRODUCTCIPPRIORITY='$ref->{PRODUCTCIPPRIORITY}', PRODUCTSTATUS='$productStatus',
             PRODUCTFILENAMEPATTERN='$ref->{PRODUCTFILENAMEPATTERN}', PRODUCTHOMEDIRECTORY='$ref->{PRODUCTHOMEDIRECTORY}',
              PRODUCTIDPSRETRANSMITLIMIT='$ref->{PRODUCTIDPSRETRANSMITLIMIT}', PRODUCTAREA='$ref->{PRODUCTAREA}',
               PRODUCTESTAVGFILESIZE='$ref->{PRODUCTESTAVGFILESIZE}', PRODUCTARCHIVE=$productArchive,
                 PRODUCTRETENTIONPERIODHR='$ref->{PRODUCTRETENTIONPERIODHR}', PRODUCTHORIZONTALRESOLUTION='$ref->{PRODUCTHORIZONTALRESOLUTION}',
                  PRODUCTVERTICALRESOLUTION='$ref->{PRODUCTVERTICALRESOLUTION}', PRODUCTFILEFORMAT='$ref->{PRODUCTFILEFORMAT}',
                   PRODUCT_TIME_COVERAGE='$ref->{PRODUCT_TIME_COVERAGE}', PRODUCT_MAP_PROJECTION='$ref->{PRODUCT_MAP_PROJECTION}',
                    PRODUCTFILEMETADATAEXTERNAL=$ref->{PRODUCTFILEMETADATAEXTERNAL}, PRODUCTFILENAMEEXTENSION='$ref->{PRODUCTFILENAMEEXTENSION}',
                     PRODUCTFILENAMEMETAEXTENSION='$ref->{PRODUCTFILENAMEMETAEXTENSION}', PRODUCTSUBTYPE='$ref->{PRODUCTSUBTYPE}',
                      INGESTINCOMINGDIRECTORYID=$incomingDirID, PRODUCTGROUPID=$productGroupID, PRODUCTCOVERAGEGAPINTERVAL_DS=$pcgi_ds,
                       PRODUCTOBSTIMEPATTERN='$ref->{PRODUCTOBSTIMEPATTERN}', PRODUCTIRMESSAGEEXTENSION='$ref->{PRODUCTIRMESSAGEEXTENSION}',
                        PRODUCTDISTFLAG='$productDistFlag'
         where PRODUCTID = $productID";

$sth = runSQL( $dbh, $sql );

### Insert into tailoring tables
if ( $update eq "n" ) {
  if( $ref->{PRODUCTFILENAMEEXTENSION} eq "h5"){
    $dbh->do( "insert into HDF5_STRUCTURE values ($productID) ");
  }
  elsif( $ref->{PRODUCTFILENAMEEXTENSION} eq "nc" ){
    $dbh->do( "insert into NETCDF4_STRUCTURE values ($productID) ");
  }
}

### Do Product Quality Summarys
#$dbh->do( "delete from PRODUCTQUALITYSUMMARY where productID=$productID" ), if ( $update eq "y" );
logger( "INFO Adding product quality summaries ..." );
my $pqsCount = 0;
for my $parm (@{$ref->{ProductQualitySummarys}{ProductQualitySummary}}) {
        logger( "INFO $parm->{PRODUCTQUALITYSUMMARYNAME} " );
        if ($update eq "n") {
        	my $sql = "insert into PRODUCTQUALITYSUMMARY
                                values ($productID, '$parm->{PRODUCTQUALITYSUMMARYNAME}',
                                '$parm->{PRODUCTQUALITYSUMMARYTYPE}', '$parm->{PRODUCTQUALITYDESCRIPTION}') ";
            $pqsCount += $dbh->do($sql);
            logger( "Added $parm->{PRODUCTQUALITYSUMMARYNAME} " );
        } else {
        	my $sql = "select PRODUCTQUALITYSUMMARYNAME from PRODUCTQUALITYSUMMARY where PRODUCTID=$productID
        				and PRODUCTQUALITYSUMMARYNAME='$parm->{PRODUCTQUALITYSUMMARYNAME}'";
        	$sth = runSQL( $dbh, $sql );   	
			my $summary=$sth->fetchrow_array();
			if (!defined($summary)){
				my $sql = "insert into PRODUCTQUALITYSUMMARY
                                values ($productID, '$parm->{PRODUCTQUALITYSUMMARYNAME}',
                                '$parm->{PRODUCTQUALITYSUMMARYTYPE}', '$parm->{PRODUCTQUALITYDESCRIPTION}') ";
            	$pqsCount += $dbh->do($sql);
            	logger( "Added $parm->{PRODUCTQUALITYSUMMARYNAME} " );
			}
        }
}


### Do Ingest Process Steps
$dbh->do( "delete from ingestprocessstep where productID=$productID" ), if ( $update eq "y" );
logger( "INFO Adding ingest process steps ..." );
my $ipsCount = 0;
for my $ips (@{$ref->{IngestProcessSteps}{IngestProcessStep}}) {

        # Check to see if this NSF exists
        my $nsfname = $ips->{NSFDESCRIPTION};
        $sth = runSQL( $dbh, "select NSFID from NDE_SUPPORTFUNCTION where NSFDESCRIPTION='$nsfname'") ;
        my $nsfid = eval { $sth->fetchrow_array()};
        if ($nsfid eq "" ) {
           logger( "ERROR NSFDescription $nsfname does not exist. Exiting..." );
           $dbh->rollback; exit(1);
        }

        # get ips options
        $optparam = $ips->{IPSOPTIONALPARAMETERS};
        $ipsfails = $ips->{IPSFAILSINGEST};
        $ipsenable = $ips->{IPSENABLE};
        $ipsretran = $ips->{IPSDORETRANSMIT};
        if( $ipsfails eq "" ){ $dbh->rollback; logger( "ERROR Missing \"IPSFAILSINGEST\" flag for IPS: $nsfname, please add!" ); exit(1);}
        if( $ipsenable eq "" ){ $dbh->rollback; logger( "ERROR Missing \"IPSENABLE\" flag for IPS: $nsfname, please add!" ); exit(1);}
        if( $ipsretran eq "" ){ $dbh->rollback; logger( "ERROR Missing \"IPSDORETRANSMIT\" flag for IPS: $nsfname, please add!" ); exit(1);}

        $sql="insert into INGESTPROCESSSTEP (PRODUCTID,NSFID, IPSOPTIONALPARAMETERS,IPSFAILSINGEST,IPSENABLE,IPSDORETRANSMIT)
                   values ($productID, $nsfid, '$optparam', '$ipsfails', '$ipsenable', '$ipsretran') ";
        $ipsCount += $dbh->do($sql);
        logger( "INFO $nsfname" );
                if( $optparam ne "" ){ logger( "INFO Optional Param: $optparam" ); }
}


### Do Product Data Source, if applicable
if ( $ref->{HOSTNAME} ne "" ) {
        logger( "INFO Adding external product data source ..." );
        $sth = runSQL( $dbh, "select HOSTID from EXTERNALDATAHOST where HOSTNAME = '$ref->{HOSTNAME}'" );
        my ( $hostID ) = $sth->fetchrow_array();
        if( $hostID eq ""){
                logger( "ERROR External data host \"$ref->{HOSTNAME}\" not found in the database. Exiting..." );
                $dbh->rollback; exit 1;
        }
        $dbh->do( "delete from PRODUCTDATASOURCE where productID=$productID and HOSTID=$hostID" ), if ( $update eq "y" );
        my $sql = "insert into PRODUCTDATASOURCE
                                        (PRODUCTID, HOSTID, PRODUCTPROVIDERDIRECTORY, PRODUCTPOLLINGFREQUENCY, PRODUCTPROVIDERPOLLINGRETRIES )
                                values ( $productID, $hostID, '$ref->{PRODUCTPROVIDERDIRECTORY}',
                                        $ref->{PRODUCTPOLLINGFREQUENCY}, $ref->{PRODUCTPROVIDERPOLLINGRETRIES} ) ";
        $sth = $dbh->do($sql);
        logger( "INFO host: $ref->{HOSTNAME}" );
}
else{
        logger( "INFO No external product data source found, skipping" );
}

### Do Platforms
$dbh->do( "delete from PRODUCTPLATFORM_XREF where productID=$productID" ), if ( $update eq "y" );
logger( "INFO Adding product platform cross-reference(s) ..." );
for my $parm (@{$ref->{Platforms}{Platform}}) {
        logger( "INFO $parm->{Platformname} " );
        $sth = runSQL( $dbh, "select PLATFORMID from PLATFORM where PLATFORMNAME = '$parm->{Platformname}'" );
        my ( $platformID ) = $sth->fetchrow_array();
        if ( defined($platformID) ){
          my $sql = "insert into PRODUCTPLATFORM_XREF values ( $platformID, $productID ) ";
          $sth = $dbh->do($sql);
        } else {
          logger( "WARN No Platformname: $parm->{Platformname} defined, not found in db." );
        }
}

$dbh->commit;


my $action = "added";
$action = "updated", if ( $update eq "y" );
logger( "INFO $psn $action successfully with $pqsCount Quality Flags and $ipsCount Ingest Process Step(s)" );

my $rc = system("cp $opt_f /opt/data/nde/$opt_m\/xmls");
if ( $rc ) {
  logger( "ERROR Failed to copy xml file: $opt_f" );
} else {
  logger( "INFO $opt_f successfully copied");
}

logger( "INFO Execution complete." );

sub logger() {
  my ( $msg ) = @_;
  print "$msg\n";
  chomp( $HMS = `date +%H:%M:%S,%N` );
  print LOG substr($HMS,0,12) . " $msg\n";

}

exit;
